REGRESSION(282990@main) LegacyEME fails when appending media segment before completing key exchange

jernoble opened the pull request:
#### 710c98ac50e2a103b0f5fea4cbb2c1179a25d986
<pre>
REGRESSION(282990@main) LegacyEME fails when appending media segment before completing key exchange
<a href="https://bugs.webkit.org/show_bug.cgi?id=282252">https://bugs.webkit.org/show_bug.cgi?id=282252</a>
<a href="https://rdar.apple.com/138002160">rdar://138002160</a>

Reviewed by Andy Estes.

When enabling the &quot;modern AVContentKey support&quot; path, we updated the modern EME implementation
but neglected to update the LegacyEME implementation with similar changes. To support the new
AVContentKeySession path, we must _not_ add the AVStreamDataParser to the AVContentKeySession.
Additionally, when keys are added through the LegacyEME path, we must explicitly signal the
SourceBufferPrivate to attempt to enqueue the protected samples for decode and display.

* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-muxed-nowait-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-muxed-nowait.html: Added.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
(WebCore::CDMSessionMediaSourceAVFObjC::addSourceBuffer):
(WebCore::CDMSessionMediaSourceAVFObjC::removeSourceBuffer):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::keyAdded):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::keyAdded):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::setCDMSession):
(WebCore::SourceBufferPrivateAVFObjC::attemptToDecrypt):

Canonical link: <a href="https://commits.webkit.org/286068@main">https://commits.webkit.org/286068@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c4230148af67e34b716f922b7a71e039345cb8d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74736 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27553 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79158 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25974 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1951 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58702 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16990 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64225 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39096 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21722 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24307 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67289 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22067 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80649 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2054 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1217 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66961 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2202 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66252 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16451 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10200 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8352 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2019 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4806 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2047 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/2967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -1365,6 +1365,13 @@ void getSupportedTypes(HashSet<String>& types) const final
+
+void MediaPlayerPrivateMediaSourceAVFObjC::keyAdded()
+{
+    if (m_mediaSourcePrivate)
+        m_mediaSourcePrivate->keyAdded();
```
```suggestion
    if (RefPtr mediaSourcePrivate = m_mediaSourcePrivate)
        mediaSourcePrivate->keyAdded();
```

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -707,17 +708,21 @@ static inline bool shouldAddContentKeyRecipients()
-    auto instanceSession = m_cdmInstance->sessionForKeyIDs(m_keyIDs);
-    if (!instanceSession)
+    if (m_cdmInstance) {
+        auto instanceSession = m_cdmInstance->sessionForKeyIDs(m_keyIDs);
```
```suggestion
        RefPtr instanceSession = m_cdmInstance->sessionForKeyIDs(m_keyIDs);
```

jyavenard commented on LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-muxed-nowait.html:
```diff
@@ -0,0 +1,67 @@
+            mimeType: 'video/mp4'
+        });
+
+        let mediaSource = new MediaSource;
```
If the test used ManagedMSE this test could be tested on iPhone

jyavenard commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -2445,3 +2445,6 @@ webanimations/scroll-timeline-clamped-current-time-while-rubber-banding-mac.html
 # webkit.org/b/281765 PROGRESSION (285251@main): [ macOS iOS ] imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect.html is a constant failure. 
 imported/w3c/web-platform-tests/navigation-api/navigate-event/cross-origin-traversal-redirect.html [ Skip ]
+
+# Modern AVContentKeySession is only available on Sonoma
```
from Sonoma?

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
```diff
@@ -68,8 +68,9 @@ ALLOW_NEW_API_WITHOUT_GUARDS_BEGIN
-    void addSourceBuffer(SourceBufferPrivateAVFObjC*);
-    void removeSourceBuffer(SourceBufferPrivateAVFObjC*);
+    enum class ShouldAddParser : bool { No, Yes };
+    void addSourceBuffer(SourceBufferPrivateAVFObjC*, ShouldAddParser);
```
this code path (existing) isn't particularly great. We have the SourceBuffer calling into the CDMSession only to have the CDMSession make a re-entrant call into the SourceBuffer to get the AVStreamDataParser. The SourceBuffer knows the information to start.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -638,13 +638,14 @@ static inline bool shouldAddContentKeyRecipients()
+    auto shouldAddParser = shouldAddContentKeyRecipients() ? CDMSessionMediaSourceAVFObjC::ShouldAddParser::Yes : CDMSessionMediaSourceAVFObjC::ShouldAddParser::No;
     if (m_session)
-        m_session->removeSourceBuffer(this);
+        m_session->removeSourceBuffer(this, shouldAddParser);
```
Wouldn't it be best to have the SourceBuffer do the work of adding the parser instead? this makes the CDMSession immediately calls back into the SourceBuffer fetch what should be a private member of the SourceBuffer

The SourceBuffer has the information on if its parser should be added or not.

jernoble committed:
REGRESSION(282990@main) LegacyEME fails when appending media segment before completing key exchange
https://bugs.webkit.org/show_bug.cgi?id=282252
rdar://138002160

Reviewed by Andy Estes.

When enabling the "modern AVContentKey support" path, we updated the modern EME implementation
but neglected to update the LegacyEME implementation with similar changes. To support the new
AVContentKeySession path, we must _not_ add the AVStreamDataParser to the AVContentKeySession.
Additionally, when keys are added through the LegacyEME path, we must explicitly signal the
SourceBufferPrivate to attempt to enqueue the protected samples for decode and display.

* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-muxed-nowait-expected.txt: Added.
* LayoutTests/http/tests/media/fairplay/legacy-fairplay-mse-muxed-nowait.html: Added.
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/CDMSessionMediaSourceAVFObjC.mm:
(WebCore::CDMSessionMediaSourceAVFObjC::addSourceBuffer):
(WebCore::CDMSessionMediaSourceAVFObjC::removeSourceBuffer):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::keyAdded):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::keyAdded):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::setCDMSession):
(WebCore::SourceBufferPrivateAVFObjC::attemptToDecrypt):

Canonical link: https://commits.webkit.org/286068@main
